src/vcpkg/vcpkgpaths.cpp

```diff
@@ -1057,45 +1057,53 @@ namespace vcpkg
         const auto tsv = prebuildinfo.platform_toolset.get();
+        const auto tsvf = prebuildinfo.platform_toolset_version.get();
         auto vsp = prebuildinfo.visual_studio_path.get();
         if (!vsp && !m_pimpl->default_vs_path.empty())
         {
             vsp = &m_pimpl->default_vs_path;
         }
+        
+        std::string error_message = "Could not find any Visual Studio instance";
 
-        if (tsv && vsp)
+        if (vsp)
         {
-            Util::erase_remove_if(
-                candidates, [&](const Toolset* t) { return *tsv != t->version || *vsp != t->visual_studio_root_path; });
-            Checks::check_exit(VCPKG_LINE_INFO,
-                               !candidates.empty(),
-                               "Could not find Visual Studio instance at %s with %s toolset.",
-                               *vsp,
-                               *tsv);
-
-            Checks::check_exit(VCPKG_LINE_INFO, candidates.size() == 1);
-            return *candidates.back();
+            Util::erase_remove_if(candidates, [&](const Toolset* t) { return *vsp != t->visual_studio_root_path; });
+            error_message += " at " + (*vsp).native();
         }
 
         if (tsv)
         {
             Util::erase_remove_if(candidates, [&](const Toolset* t) { return *tsv != t->version; });
-            Checks::check_exit(
-                VCPKG_LINE_INFO, !candidates.empty(), "Could not find Visual Studio instance with %s toolset.", *tsv);
+            error_message += " with " + *tsv + " toolset";
         }
 
-        if (vsp)
+        if (tsvf)
         {
-            const auto vs_root_path = *vsp;
-            Util::erase_remove_if(candidates,
-                                  [&](const Toolset* t) { return vs_root_path != t->visual_studio_root_path; });
-            Checks::check_exit(
-                VCPKG_LINE_INFO, !candidates.empty(), "Could not find Visual Studio instance at %s.", vs_root_path);
+            Util::erase_remove_if(candidates, [&](const Toolset* t) {
+                const auto requested_version = *tsvf;
+                if (requested_version == t->full_version)
+                {
+                    return false;
+                }
+                // Check if requested version is part of the full version, ie "14.25" should match "14.25.28610"
+                return !(requested_version.size() < t->full_version.size() &&
+                         t->full_version[requested_version.size()] == '.' &&
+                         t->full_version.substr(0, tsvf->size()) == requested_version);
+            });
+            error_message += " for toolset version " + *tsvf;
         }
 
-        Checks::check_exit(VCPKG_LINE_INFO, !candidates.empty(), "No suitable Visual Studio instances were found");
-        return *candidates.front();
+        Checks::check_exit(VCPKG_LINE_INFO, !candidates.empty(), error_message + ".");
 
+        if (tsv && vsp)
+        {
+            return *candidates.back();
```

Comment:
This seems rather odd behavior – why would we want the oldest available version in this case, instead of the newest? This line does not introduce this behavior, but preserves it as it was before.

Reply:
Ah, I see, I missed that. What I don't see is why that assert should be here. And it isn't on my machine:
```
C:\Program Files (x86)\Microsoft Visual Studio\2019\Professional\VC\Tools\MSVC>dir /B /W
14.22.27905
14.25.28610
14.29.30037
```

Custom toolchain file `x64-windows-test`:
```
set(VCPKG_VISUAL_STUDIO_PATH [[C:\Program Files (x86)\Microsoft Visual Studio\2019\Professional]])
set(VCPKG_PLATFORM_TOOLSET v142)
```

master vcpkg:
```
$ ./vcpkg.exe install zlib:x64-windows-test
Computing installation plan...
The following packages will be built and installed:
    zlib[core]:x64-windows-ts -> 1.2.11#12
Error: vcpkg has crashed; no additional details are available.
The source line is D:\a\_work\1\s\src\vcpkg\vcpkgpaths.cpp(1076)
```

I assume it's fine if I don't add that crash back in. I'll just use `front()` in this case.